File: Switches/Service/DataCollectorServiceSwitches.cs
using IDAUtil;
using IDAUtil.Model.Properties.ServerProperty;
using IDAUtil.Model.Properties.TaskProperty;
using IDAUtil.Model.Properties.TcodeProperty.CO09Obj;
using IDAUtil.Model.Properties.TcodeProperty.ZV04Obj;
using IDAUtil.Service;
using lib;
using Microsoft.VisualBasic.CompilerServices;
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;

namespace Switches {
    public class DataCollectorServiceSwitches {
        private readonly IDataCollectorServer dataCollectorServer;
        private readonly IDataCollectorSap dataCollectorSap;

        public DataCollectorServiceSwitches(IDataCollectorServer dataCollectorServer, IDataCollectorSap dataCollectorSap) {
            this.dataCollectorServer = dataCollectorServer;
            this.dataCollectorSap = dataCollectorSap;
        }

        public List<SwitchesProperty> getSwitchesList(string salesOrg, string id) {
            var ZVList = getZvList(salesOrg);
            if (ZVList is null) { return null;}

            var cdList = getCdList(salesOrg);
            var sdList = getSdList(salesOrg);

            var allSoldToList = getInitialSwitchPropertyList(id, sdList, ZVList, cdList, isSpecifficSoldTo: false);
            var SpecifficSoldToList = getInitialSwitchPropertyList(id, sdList, ZVList, cdList, isSpecifficSoldTo: true);
            var finalList = new List<SwitchesProperty>();

            if (SpecifficSoldToList.Count > 0) {

                finalList.AddRange(SpecifficSoldToList);

                //remove all speciffic sold to switch from general switches
                allSoldToList = allSoldToList
                    .Where(x => !SpecifficSoldToList
                    .Any(y => y.soldTo == x.soldTo && y.oldSku == x.oldSku))
                    .ToList();
            }

            finalList.AddRange(allSoldToList);
            finalList = getFinalListWithStockDetails(finalList, salesOrg);
            finalList.OrderBy(x => x.soldTo).ThenBy(x => x.item);

            return finalList;
        }

        private List<CustomerDataProperty> getCdList(string salesOrg) {
            return dataCollectorServer.getCustomerDataList(salesOrg);
        }

        private List<SwitchesProperty> getFinalListWithStockDetails(List<SwitchesProperty> list, string salesOrg) {
            var sap = Create.sapLib();
            var co09 = new CO09(sap);

            var listOfUniqueOldSkus = (from x in list select x.oldSku).Distinct().ToList();
            var listOfUniqueNewSkus = (from x in list select x.newSku).Distinct().ToList();

            var co09ObjOldSkuList = new List<CO09Property>();
            var co09ObjNewSkuList = new List<CO09Property>();

            foreach (var item in listOfUniqueOldSkus) co09ObjOldSkuList.Add(co09.getStockDetails(item, salesOrg));
            foreach (var item in listOfUniqueNewSkus) co09ObjNewSkuList.Add(co09.getStockDetails(item, salesOrg));

            // assigns value to final list passed to func from both old and new sku CO09 lists
            foreach (var listItem in list) {
                foreach (var co09OldItem in co09ObjOldSkuList) {
                    if (listItem.oldSku == co09OldItem.sku) {
                        listItem.oldSkuATP = co09OldItem.ATP;
                        listItem.oldSkuRecoveryDate = co09OldItem.recoveryDate;
                        listItem.oldSkuRecoveryQty = co09OldItem.recoveryQty;
                        break;
                    }
                }

                foreach (var co09Newtem in co09ObjNewSkuList) {
                    if (listItem.newSku == co09Newtem.sku) {
                        listItem.newSkuATP = co09Newtem.ATP;
                        listItem.newSkuRecoveryDate = co09Newtem.recoveryDate;
                        listItem.newSkuRecoveryQty = co09Newtem.recoveryQty;
                        break;
                    }
                }
            }
            return list;
        }

        private List<SwitchesProperty> getInitialSwitchPropertyList(string id, List<SwitchesDataProperty> sdList, List<ZV04IProperty> zvList, List<CustomerDataProperty> cdList, bool isSpecifficSoldTo) {
            return (
                    from zv in zvList
                    join cd in cdList on new { key0 = zv.soldTo, key1 = zv.shipTo } equals new { key0 = cd.soldTo, key1 = cd.shipTo }
                    join sd in sdList on new { key0 = zv.material, key1 = cd.country.ToLower() } equals new { key0 = sd.oldSku, key1 = sd.country.ToLower() }
                    where getStockConditon(sd, zv) && getSoldToCondition(isSpecifficSoldTo, zv, sd)
                    select getSwitchPropertyObjs(zv, sd, cd, id)
                    ).ToList();
        }

        private static bool getSoldToCondition(bool isSpecifficSoldTo, ZV04IProperty zv, SwitchesDataProperty sd) {
            return isSpecifficSoldTo ? sd.soldTo == zv.soldTo : sd.soldTo == 0;
        }

        private List<SwitchesDataProperty> getSdList(string salesOrg) {
            return dataCollectorServer.getSwitchesDataList(salesOrg).Where(x => DateTime.Today >= Conversions.ToDate(x.startDate) && DateTime.Today <= Conversions.ToDate(x.endDate)).ToList();
        }

        private List<ZV04IProperty> getZvList(string salesOrg) {
            var list = dataCollectorSap.getZV04IList(salesOrg, IDAEnum.Task.switches);

            if (list is null) {
                return null;
            } else {
                return list.Where(x => (x.delBlock ?? "") != IDAConsts.DelBlocks.leadTimeBlock && (x.delBlock ?? "") != "Z4" && (x.delBlock ?? "") != "04" && string.IsNullOrEmpty(x.rejReason)).ToList();
            }
        }

        private SwitchesProperty getSwitchPropertyObjs(ZV04IProperty zv, SwitchesDataProperty sd, CustomerDataProperty cd, string id) {
            return new SwitchesProperty(salesOrg: cd.salesOrg,
                                        country: cd.country,
                                        soldTo: zv.soldTo,
                                        switchForCustomer: sd.soldTo == 0 ? "All" : "Speciffic",
                                        shipTo: zv.shipTo,
                                        order: zv.order,
                                        shipToName: zv.shipToName,
                                        item: zv.item,
                                        orderedQty: zv.orderQty,
                                        confirmedQty: zv.confirmedQty,
                                        oldSku: sd.oldSku,
                                        oldSkuDescription: sd.oldSkuDescription,
                                        oldSkuCaseBarcode: sd.oldSkuCaseBarcode,
                                        oldSkuUnitBarcode: sd.oldSkuUnitBarcode,
                                        oldSkuATP: default,
                                        oldSkuRecoveryDate: default,
                                        oldSkuRecoveryQty: default,
                                        newSku: sd.newSku,
                                        newSkuDescription: sd.newSkuDescription,
                                        newSkuCaseBarcode: sd.newSkuCaseBarcode,
                                        newSkuUnitBarcode: sd.newSkuUnitBarcode,
                                        newSkuATP: default,
                                        newSkuRecoveryDate: default,
                                        newSkuRecoveryQty: default,
                                        startDate: Conversions.ToDate(sd.startDate),
                                        endDate: Conversions.ToDate(sd.endDate),
                                        needOutOfStockToSwitch: sd.needOutOfStockToSwitch,
                                        switchAutomatic: sd.switchAutomatic,
                                        switchComment: !string.IsNullOrEmpty(sd.switchComment) ? sd.switchComment.Replace("'","") : "" ,
                                        id: id);
        }

        private bool getStockConditon(SwitchesDataProperty sd, ZV04IProperty zv) {
            return (sd.needOutOfStockToSwitch == true && zv.confirmedQty < zv.orderQty) || sd.needOutOfStockToSwitch == false;
        }
    }
}

File: Switches/Service/DataCollectorServiceSwitches.vb
Imports IDAUtil
Imports [lib]

Public Class DataCollectorServiceSwitches
    Private ReadOnly dataCollectorServer As IDataCollectorServer
    Private ReadOnly dataCollectorSap As IDataCollectorSap

    Public Sub New(dataCollectorServer As IDataCollectorServer, dataCollectorSap As IDataCollectorSap)
        Me.dataCollectorServer = dataCollectorServer
        Me.dataCollectorSap = dataCollectorSap
    End Sub

    Public Function getSwitchesList(salesOrg As String, id As String) As List(Of SwitchesProperty)
        Dim ZVList As List(Of ZV04IProperty) = getZvList(salesOrg)
        If ZVList Is Nothing Then Return Nothing
        Dim cdList As List(Of CustomerDataProperty) = GetCdList(salesOrg)
        Dim sdList As List(Of SwitchesDataProperty) = getSdList(salesOrg)

        Dim allSoldToList As List(Of SwitchesProperty) = GetLinq(id, sdList, ZVList, cdList, isSpecifficSoldTo:=False)
        Dim SpecifficSoldToList As List(Of SwitchesProperty) = GetLinq(id, sdList, ZVList, cdList, isSpecifficSoldTo:=True)
        Dim finalList As New List(Of SwitchesProperty)

        If SpecifficSoldToList.Count > 0 Then
            finalList.AddRange(SpecifficSoldToList)

            'removes speciffic sold to switches from general switches for all sold tos
            allSoldToList = allSoldToList.Where(Function(x) SpecifficSoldToList.Any(Function(y) y.soldTo <> x.soldTo)).ToList
        End If

        finalList.AddRange(allSoldToList)

        finalList = getFinalListWithStockDetails(finalList, salesOrg)

        finalList.OrderBy(Function(x) x.soldTo).ThenBy(Function(x) x.item)

        Return finalList
    End Function

    Private Function GetCdList(salesOrg As String) As List(Of CustomerDataProperty)
        Return dataCollectorServer.getCustomerDataList(salesOrg)
    End Function

    Private Function getFinalListWithStockDetails(list As List(Of SwitchesProperty), salesOrg As String) As List(Of SwitchesProperty)
        Dim sap As ISAPLib = Create.sapLib
        Dim co09 As New CO09(sap)

        Dim listOfUniqueOldSkus As List(Of Long) = (From x In list Select x.oldSku).Distinct.ToList
        Dim listOfUniqueNewSkus As List(Of Long) = (From x In list Select x.newSku).Distinct.ToList

        Dim co09ObjOldSkuList As New List(Of CO09Property)
        Dim co09ObjNewSkuList As New List(Of CO09Property)

        For Each item In listOfUniqueOldSkus
            co09ObjOldSkuList.Add(co09.getStockDetails(item, salesOrg))
        Next

        For Each item In listOfUniqueNewSkus
            co09ObjNewSkuList.Add(co09.getStockDetails(item, salesOrg))
        Next

        'assigns value to final list passed to func from both old and new sku CO09 lists
        For Each listItem In list
            For Each co09OldItem In co09ObjOldSkuList
                If listItem.oldSku = co09OldItem.sku Then
                    listItem.oldSkuATP = co09OldItem.ATP
                    listItem.oldSkuRecoveryDate = co09OldItem.recoveryDate
                    listItem.oldSkuRecoveryQty = co09OldItem.recoveryQty
                    Exit For
                End If
            Next

            For Each co09Newtem In co09ObjNewSkuList
                If listItem.newSku = co09Newtem.sku Then
                    listItem.newSkuATP = co09Newtem.ATP
                    listItem.newSkuRecoveryDate = co09Newtem.recoveryDate
                    listItem.newSkuRecoveryQty = co09Newtem.recoveryQty
                    Exit For
                End If
            Next
        Next

        Return list
    End Function

    Private Function GetLinq(id As String, sdList As List(Of SwitchesDataProperty), zvList As List(Of ZV04IProperty), cdList As List(Of CustomerDataProperty), isSpecifficSoldTo As Boolean) As List(Of SwitchesProperty)
        Return (From zv In zvList
                Join cd In cdList
                    On zv.soldTo Equals cd.soldTo And zv.shipTo Equals cd.shipTo
                Join sd In sdList
                   On zv.material Equals sd.oldSku And cd.country.ToLower Equals sd.country.ToLower
                Where getStockConditon(sd, zv) AndAlso getSoldToCondition(isSpecifficSoldTo, zv, sd)
                Select getSwitchPropertyObjs(zv, sd, cd, id)).ToList()
    End Function

    Private Shared Function getSoldToCondition(isSpecifficSoldTo As Boolean, zv As ZV04IProperty, sd As SwitchesDataProperty) As Boolean
        Return If(isSpecifficSoldTo, sd.soldTo = zv.soldTo, sd.soldTo = 0)
    End Function

    Private Function getSdList(salesOrg As String) As List(Of SwitchesDataProperty)
        Return dataCollectorServer.getSwitchesDataList(salesOrg).Where(Function(x) Date.Today >= x.startDate AndAlso Date.Today <= x.endDate).ToList
    End Function

    Private Function getZvList(salesOrg As String) As List(Of ZV04IProperty)
        Dim list As List(Of ZV04IProperty) = dataCollectorSap.getZV04IList(salesOrg, IDAEnum.Task.switches)
        If list Is Nothing Then
            Return Nothing
        Else
            Return list.Where(Function(x) x.delBlock <> IDAConsts.bypassOrderChangesBlock AndAlso x.delBlock <> "Z4" AndAlso x.delBlock <> "04" AndAlso x.rejReason = "").ToList
        End If
    End Function

    Private Function getSwitchPropertyObjs(zv As ZV04IProperty, sd As SwitchesDataProperty, cd As CustomerDataProperty, id As String) As SwitchesProperty
        Return New SwitchesProperty(cd.salesOrg, cd.country, zv.soldTo, If(sd.soldTo = 0, "All", "Speciffic"), zv.shipTo, zv.order, zv.shipToName, zv.item, zv.orderQty, zv.confirmedQty, sd.oldSku, sd.oldSkuCaseBarcode, sd.oldSkuUnitBarcode, Nothing, Nothing, Nothing, sd.newSku, sd.newSkuCaseBarcode, sd.newSkuUnitBarcode, Nothing, Nothing, Nothing, sd.startDate, sd.endDate, sd.needOutOfStockToSwitch, sd.switchAutomatic, sd.switchComment, id)
    End Function

    Private Function getStockConditon(sd As SwitchesDataProperty, zv As ZV04IProperty) As Boolean
        Return (sd.needOutOfStockToSwitch = True AndAlso zv.confirmedQty < zv.orderQty) OrElse sd.needOutOfStockToSwitch = False
    End Function
End Class
